### PR TITLE
Tutorial Playground

### PR DESCRIPTION
**Pull request by CodaFi**

### CODE REVIEW

File: Tutorial.playground/Contents.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//: * [Polymorphism](https://en.wikipedia.org/wiki/Polymorphism)
+//: * [Map/Filter/Reduce](https://medium.com/@ivicamil/higher-order-functions-in-swift-part-1-d8e75f963d13)
+//:
+//: What this tutorial does *not* require are that you know
+//:
+//: * [Abstract Nonsense](https://en.wikipedia.org/wiki/Abstract_nonsense)
+//: * Functional Programming
+//: * What that dumb 'M' word is/does/means
+
+//: # Introduction
```
Comment (CodaFi) [inline Tutorial.playground/Contents.swift]: @rnapier I know you stressed pre-reqs in the last pull request.  Is there anything you would add/remove from this list?  Any better links than wikipedia would be helpful too.

**Comment by zbeckman** [comment]
I'm trying to get SwiftCheck working with DateTime (https://github.com/zbeckman/DateTimeKit). I get that I need to create an `extension DateTime : Arbitrary` but I'm a bit lost as to how to do this right. If I want SwiftCheck to support DateTime just as thoroughly as an atomic type, how would I go about that?

**Comment by CodaFi** [comment]
NSDate cannot be used directly (dumbass Self requirement restrictions on protocol instances).  You have to wrap it.

**Comment by CodaFi** [comment]
There, I included it in the tutorial.  It's a pretty great way to introduce modifiers.

**Comment by zbeckman** [comment]
Excellent. That's another major real world problem we run into all the time. (Also the fact that `NSDate` kinda sucks which is why we are using DateTime instead...)

File: Tutorial.playground/Contents.swift [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+
+//: Finally, the RFC says the TLD for the address can only consist of lowercase letters with a length larger than 1.
+
+let tld = lowerCaseLetters.proliferateNonEmpty().suchThat({ $0[$0.endIndex.predecessor()] != "." }).fmap(String.init)
+
+//: So now we've got all the pieces together, so how do we put them together to make the final generator?  Well, how
+//: about some glue?
+
+// Concatenates 5 strings together in order.
+func glue5(l : String) -> String -> String -> String -> String -> String {
```
Comment (jtbandes) [inline Tutorial.playground/Contents.swift]: This should work too:

    func glue5(l: String)(m: String)(m2: String)(m3: String)(r: String) {
        return l + m + m2 + m3 + r
    }

Comment (CodaFi) [inline Tutorial.playground/Contents.swift]: Absolutely.  I'll use that.

File: Tutorial.playground/Contents.swift [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+generatorBoundedSizeArrays_.generate
+generatorBoundedSizeArrays_.generate
+generatorBoundedSizeArrays_.generate
+
+//: Now that you've seen what generators can do, we'll use all we've learned to create a generator that
+//: produces email addresses.  To do this, we'll need one more operator/method notated `<*>` or `ap`.
+//: `ap` comes from [Applicative Functors](http://staff.city.ac.uk/~ross/papers/Applicative.html) and is
+//: used to "zip together" `Gen`erators of functions with `Gen`erators of of values, applying each function
+//: during the zipping phase.  That definition is a little hand-wavey and technical, so for now we'll say that
+//: `ap` works like "glue" that sticks special kinds of generators togethers.
```
Comment (CodaFi) [inline Tutorial.playground/Contents.swift]: This is a dumb and lazy and inaccurate simplification.  I think I'll use a diagram.

File: Tutorial.playground/Contents.swift [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+//: # All Together Now!
+
+//: If you've made it this far, congratulations!  That's it.  Naturally, there are other combinators
+//: and fancy ways of creating `Gen`erators and properties with the primitives in this framework, 
+//: but they are all variations on the themes present in ths tutorial.  With the power of SwiftCheck
+//: and a sufficiently expressive testing suite, we can begin to check our programs not for 
+//: individual passing cases in a few scattershot unit tests, but declare and enforce immutable 
+//: properties that better describe the intent and invariants of our programs.  If you would like 
+//: further reading, see the files `Arbitrary.swift`, `Test.swift`, `Modifiers.swift`, and 
+//: `Property.swift`.
```
Comment (CodaFi) [inline Tutorial.playground/Contents.swift]: I'd like a list of external links here too

**Comment by rnapier** [comment]
I like where this is going, but I'm struggling to use it in my own code. You have a lot of nice examples like `emailGen`, but don't show how you would use that with `forAll` (or more generally, how to use it with `property`). With experimentation, I think I've discovered that these can't be passed to `forAll` and have to instead be generated inside the closure. It's not clear whether this is correct or not.

**Comment by CodaFi** [comment]
I can see 2 ways of going about your specific example:

1) Pass the generator to `forAll`

```swift
let arrayGen = UInt8.arbitrary.proliferateSized(5)

property("...") <- forAll(arrayGen) { x in
    /// ... 
}
```

2) Use implication to discard bad arrays:

```swift
property("") <- forAll { (x : [UInt8]) in
    return (x.count == 8) ==> {
        /// ...
    }
}
```

The former is more conducive to explanation than the latter.

File: SwiftCheck/Gen.swift [hunk 5]
```diff
@@ -33,10 +33,8 @@ public struct Gen<A> {
 	/// only that value.
 	///
 	/// The input collection is required to be non-empty.
-	public static func fromElementsOf<S : CollectionType where S.Generator.Element == A, S.Index : protocol<RandomType, BidirectionalIndexType>>(xs : S) -> Gen<A> {
-		assert(!xs.isEmpty, "Gen.fromElementsOf used with empty sequence")
-
-		return choose((xs.startIndex, xs.endIndex.predecessor())).fmap { i in
+	public static func fromElementsOf<S : Indexable where S.Index : protocol<Comparable, RandomType>>(xs : S) -> Gen<S._Element> {
```
Comment (CodaFi) [inline SwiftCheck/Gen.swift]: This evil is only temporary.

**Comment by rnapier** [comment]
Does the filtering (implication) process at up my 100 tests? I mean, if I pick 100 arrays and only 2 of them are the right length, do I run just those 2, or do I keep implicating until get I get 100 things to try?

I didn't realize the `forAll(generator)...` form. That probably answers my next question about how to create non-empty strings.

**Comment by rnapier** [comment]
Hmm, I feel so close, but I'm breaking the compiler (I expect this is getting off-topic for this PR, but it's a handy place to explore the question, and the kind of question that popped up as soon as I started trying to use this).

        let passwordGenerator = String.arbitrary.suchThat { $0 != "" }
        let saltGenerator = UInt8.arbitrary.proliferate()
        let gen = Gen.zip(passwordGenerator, saltGenerator)

This fails to compile with:

    Cannot invoke 'zip' with an argument list of type '(Gen<String>, Gen<[UInt8]>)'
        Expected an argument list of type '(Gen<A>, Gen<B>)'

That is clearly the wrong error message, but suggests I'm doing *something* wrong.

**Comment by CodaFi** [comment]
Oh, yeah.  You notice how I use explicit specifications with Gen everywhere?  It's more than just preference, it's often required.  Besides, it serves as an excellent marker of intent that can be checked by the compiler -bonus.

**Comment by CodaFi** [comment]
OK, I believe this is the structure I'm going to stick with.  Modifications and suggestions should be narrower from here on out.

**Comment by zbeckman** [comment]
Really, really want to know how to use a callback. Trying something like this:
```swift
it("generates a large list of unique numbers") {
	let numericGen: Gen<Int> = Int.arbitrary
	var listOfNumbers: [Int] = []
					
	property("numbers are unique") <- forAll(numericGen) { (i: Int) in
		return (listOfNumbers.contains(i)) ==> {
			return (listOfNumbers.count < { listOfNumbers.append(i); return listOfNumbers.count }())
		}
	}.withCallback(expect(listOfNumbers.count == 100))
}
```

**Comment by zbeckman** [comment]
Also, suggest you add something that explains the intricacies of shrinking, and perhaps give this example:
```swift
// Shrinking can introduce whitespace. Make sure those cases are excluded from the test case results.
					
return (e =~ "[\\s\\n]") ==> {
	return f.isValid
}
```

**Comment by CodaFi** [comment]
Shrinking introduces whitespace because you have failed a test case in your example.  You shouldn't be trying to avoid lesser cases that could be important while debugging your code by turning off the shrinker, you should be examining the spec to see why you failed on the first datum, or writing a Modifier type that shrinks the way you want it to. I make no guarantees about how and when the testing block will be invoked, only that at least 100 cases + shrinking will occur.  Perhaps that means that you need to convert your test to a more law-like form rather than a unit-test like form.

**Comment by CodaFi** [comment]
As for the callbacks thing, I appreciate what QuickCheck (and consequently SwiftCheck) did by making `withCallback` verbose and a little tough to use.  There's rarely a case where you need to attach an effect to a test case that you can't already do with `.verbose`, and when you do you only really care about failures, so we provide `withFail` and `withEachFail` for that.  I've already shown the former in the tutorial, I don't wish to encourage it any further.

**Comment by CodaFi** [comment]
```swift
// I make no guarantees of uniqueness.
it("generates a large list of unique numbers") {
    let numericGen: Gen<Int> = Int.arbitrary
    var listOfNumbers: [Int] = [] // Uh-oh, mutable state.

    // This is not a law, this is something that's taking advantage of implementation details
    // in the shrinking loop.
    property("numbers are unique") <- forAll(numericGen /* This is the same as just a plain forAll */) { (i: Int) in
        return (listOfNumbers.contains(i)) ==> {
            return (listOfNumbers.count < { listOfNumbers.append(i); return listOfNumbers.count }())
        }
    // You have no guarantees about how many times the testing block is run.
    }.withCallback(expect(listOfNumbers.count == 100))
}
```

To convert something like this into a law you need a specification and a function that obeys some part of that spec.  Here, the only reasonable assumption to make is that the testing block will be invoked some non-zero amount of times.  The only "law" you can extract here is this:

it("generates numbers") {
    var s : Set = Set() // Still makes me uneasy
    property("it generates numbers") <- forAll({ (x : Int) {
        s.insert(x)
        return true // What else can you do?
    })
}

Which is nigh useless.

File: Tutorial.playground/Contents.swift [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+//: rather than break the program.  Throughout this tutorial, simplifications like the above will be
+//: made to aid your understanding.  Towards the end of it, we will begin to remove much of these 
+//: "training wheels" and reveal the real concepts and types of the operations in the library, which
+//: are often much more powerful and generic than previously presented.
+//:
+//: Unlike Unit Testing, Property Testing abhors the use of state and global variables.  Property 
+//: Tests are local, atomic entities that ideally only use the data given to them to match a 
+//: user-defined specification for the behavior of a program or algorithm.  While this may
+//: seem draconian, the upshot of following these [unwritten] rules is that the produced tests become
+//: "law-like", as in a [Mathematical or Scientific Law](https://en.wikipedia.org/wiki/Laws_of_science).
```
Comment (CodaFi) [inline Tutorial.playground/Contents.swift]: @zbeckman Does this further your understanding of why your use-cases are a little strange w.r.t. this library?

**Comment by zbeckman** [comment]
Hm. I see your points.

However – let's say you wanted to write a test that ensures a given list of data is populated with at least 100 (or more?) "valid" sets of data. We could say a valid piece of data is a "positive Int" or a "well formed email" or a "password/salt pair." Doesn't really matter – one thing I run into in the real world is a two phase test plan:

1. Test that your preconditions are valid, e.g., that you have actually create 100 viable and usable properties within the definition of your spec.
2. Run those viable properties through your program to see if it fails.

The above example was a bit contrived but that's what it was getting at... Validate the properties, and THEN verify the program.

**Comment by CodaFi** [comment]
Then you would make part 1 a `Gen`erator and part 2 the actual property.

**Comment by zbeckman** [comment]
One thing to understand about how tools like SwiftCheck get applied... and I'm saying this from about 30 years in SQA, from the Defense Department to little startups... it gets applied in a very "practical" way. Teams that are working to deliver working software end up applying it to solve their immediate need. That includes things like:

1. Generate a set of test data for a given program component.
2. Validate that the test data is, actually, valid test data. Often the generation of the data set is just as complicated as testing the code or even writing the code.
3. FIGURING OUT how to test a program... because they are not mathematicians or professional quality assurance people. Or perhaps they are profession QA people... but that tends to mean they are NOT programmers.
4. Applying reproductions to ensure regressions in a UI or UI component or program component do not occur.
5. Actually GENERATING real data for the program – either in a quality assurance setting (e.g. load a QA server with a lot of test data that looks "real") or even in production (use it to populate required values in a running system – more rare, but it happens).

I realize some of these fall outside the scope of what SwiftCheck/QuickCheck was really designed to do, but when you give a programmer a tool and a problem... ;-)

**Comment by CodaFi** [comment]
I know it's a very strange thing to be told to throw out that kind of experience for this newfangled abstract nonsense, but there are very serious implications to this whole thing.  When you get down to properties, I mean *really* get down to properties, you only have 2 steps:

1) What are my assumptions?
2) When do they hold?

Step 1 lets you break down your problem into properties, your assumptions become inputs to functions.
Step 2 lets you actually write some "laws".  SwiftCheck may not be able to tell you with 100% certainty that they hold, but like a Scientific Law we take repeated observation of success to mean success.

**Comment by zbeckman** [comment]
(And yes, making a custom shrinker makes sense – however, I can pretty much guarantee that 80% of the people that use this library will be utterly confused when they write an email generator that explicitly does NOT include spaces... and then they see failure cases that do – that is the only reason I'm suggesting you go down this path – to avoid further confusion).

**Comment by CodaFi** [comment]
*Utterly*, no.  I catch your meaning tho.  I've carved out room for another example under the modifier types one.  I will solve your problem there.

**Comment by CodaFi** [comment]
There, that should suffice.

**Comment by zbeckman** [comment]
Looking back at our conversation, I have a question: How would you write a generator that would create a list of 100 things that you _know_ meet a certain program specification? (Perhaps email addresses that you know pass the `email =~ validEmailPattern` or whatever)? I'm curious to see your implementation of a generator that verifies such a rule _and_ produces a list. I think this would be another good real world case. The key here is making it clear to the read:
1. How to generate a list of things.
2. How to size the list to whatever desired size.
3. How to verify that each thing meets a program specification.
(I think you've touched on all of these but I don't think you have put it all together in one place – which makes me wonder if I'm doing it the right way).

**Comment by CodaFi** [comment]
It's more efficient to be correct by construction than to be correct by filtering out the totality of strings SwiftCheck can generate.  I wouldn't make a generator of strings with a `suchThat` clause using `~=` as a filter because Gen will spin as long as it needs trying to find a a passing example.  Convert the body of the relation `~=` into the blocks necessary to build the generator, 

```swift
let g = Gen.sized >>- { n in
    return myGen.proliferateSized(n)
} >>- Gen.fromElementsOf
```

**Comment by zbeckman** [comment]
How would I go about having one test run a larger set of property tests (say 1000 rather than 100)?

**Comment by CodaFi** [comment]
That is accomplished in QuickCheck by modifying the stdArgs options in.  I haven't explored a satisfactory way of doing the same in Swift.  If you'd like to try, all the machinery is there.

**Comment by zbeckman** [comment]
What about adding a few props to Testable (maxRun, maxFail, things like that)?

**Comment by zbeckman** [comment]
(I can give it a try but I'm no expert on this syntax).

**Comment by CodaFi** [comment]
No, don't do that.  There's an internal struct called `Argument` that is passed to `quickCheckWithResult`.  You need to make that public and probably find a way to pass it through the `property` function.

**Comment by CodaFi** [comment]
With that, I'm done.  I will absolutely accept modifications to this in the form of pull requests, but any future questions about any part of the framework need to be separate issues from here on out.

Thank you all for your help with this :two_hearts: 

:boat:

**Comment by zbeckman** [comment]
Awesome! This is a big gain the project. Question: Can you generate an HTML page from the tutorial, stick it into the project, and link to it from the readme.md so it's on the wiki? That would be great if it could be done.

By the way... my comment on adding props to Testable where motivated by ScalaCheck, where I can do this (and which I really like):
```swift
property("Unicode email addresses validate correctly") <- forAll(Generators.emailAddress) { (e : String) in
		f.text = e
		return f.isValid
}.minSize(200).maxDiscard(50)
```
Anyhow, food for future thought...

**Comment by CodaFi** [comment]
Why HTML?  The playground comes bundled with the project now, so you have access to it even if you build with carthage.  The tutorial is designed to be interactive.

As for the other stuff, like I said, [it's all there](https://github.com/typelift/SwiftCheck/blob/swift-develop/SwiftCheck/Test.swift#L243-L250), it just needs a public interface.

**Comment by zbeckman** [comment]
So that it's on the wiki. Our team doesn't get the Carthage checkout since frameworks are handled by me and then committed in git. Nothing like good docs on the web.

**Comment by CodaFi** [comment]
Do you know of any tools to extract that kind of information from a playground?

**Comment by zbeckman** [comment]
No afraid not... will keep my eyes open for something though.

**Pull request merged**